modape/modis/download.py

```diff
@@ -234,55 +272,71 @@ def download(self,
             overwrite (bool): Replace existing files.
             multithread (bool): Use multiple threads for downloading.
             nthreads (int): Number of threads.
+            max_retries (int): Maximum number of retries for failed downloads (for no max, set -1).
+            robust (bool): Perform robust downloading (checks file size and checksum).
 
         Raises:
             DownloadError: If one or more errors were encountered during downloading.
+        Returns:
+            List of downloaded MODIS HDF file IDs.
         """
 
         # make sure target directory is dir and exists
         assert targetdir.is_dir()
         assert targetdir.exists()
 
-        with SessionWithHeaderRedirection(username, password) as session:
+        retry_count = 0
+        to_download = self.results.copy()
+        downloaded = []
 
-            retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])
-            session.mount(
-                "https://",
-                HTTPAdapter(pool_connections=nthreads, pool_maxsize=nthreads*2, max_retries=retries)
-            )
+        while True:
 
-            if multithread:
-                log.debug("Multithreaded download using %s threads. Warming up connection pool.", nthreads)
-                # warm up pool
-                _ = session.get(self.results[0]["link"], stream=True, allow_redirects=True)
+            with SessionWithHeaderRedirection(username, password) as session:
 
-                with ThreadPoolExecutor(nthreads) as executor:
+                retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])
```

Comment:
I like the idea in general - but aren't we opening the door to _very long_  retries? Not so sure if that's optimal. 

Maybe we should also impose a maximum number of re-tries in the while loop - setting to `-1` leads to an infinite loop.

Reply:
Yes, let's maximise to 4 hours?

The infinite loop, as currently implemented, is a great approach: let's assume the user downloads data because he needs it... Not just for fun; what I mean to say is: MODAPE keep trying is a great feature. ;-)

Reply:
> With regard to wait time: could also code the back-off factor as:
> `backoff_factor = 2**retry_count`

4 hours divided by 32 seconds = 450; so for maximising the factor we need some code like:
`min(450, backoff_factor = 2**retry_count)`